ts/Series/Treegraph/TreegraphSeries.ts

```diff
@@ -0,0 +1,659 @@
+            }
+            if (node.xPosition >= maxX) {
+                maxX = node.xPosition;
+                maxXSize = Math.max(nodeSizeX as number, maxXSize);
+            }
+            if (node.yPosition <= minY) {
+                minY = node.yPosition;
+                minYSize = Math.max(nodeSizeY as number, minYSize);
+            }
+            if (node.yPosition >= maxY) {
+                maxY = node.yPosition;
+                maxYSize = Math.max(nodeSizeY as number, maxYSize);
+            }
+        });
+
+        // Calculate the values of linear transformation, which will later be
+        // applied as `nodePosition = a * x + b` for each direction.
+        const ay = maxY === minY ?
+                1 :
+                (plotSizeY - (minYSize + maxYSize) / 2) / (maxY - minY),
+            by = maxY === minY ? plotSizeY / 2 : -ay * minY + minYSize / 2,
+            ax = maxX === minX ?
+                1 :
+                (plotSizeX - (maxXSize + maxXSize) / 2) / (maxX - minX),
+            bx = maxX === minX ? plotSizeX / 2 : -ax * minX + minXSize / 2;
+
+        return { ax, bx, ay, by };
+    }
+
+    /* eslint-disable valid-jsdoc */
+    public translate(): void {
+        if (!this.processedXData) {
+            this.processData();
+        }
+        this.generatePoints();
+        this.nodeColumns = this.createNodeColumns();
+
+        const series = this,
+            options = this.options,
+            nodeColumns = this.nodeColumns;
+
+        this.nodePadding = this.getNodePadding();
+
+        // Calculate level options used in sankey and organization
+        series.mapOptionsToLevel = getLevelOptions({
+            // NOTE: if support for allowTraversingTree is added, then from
+            // should be the level of the root node.
+            from: 1,
+            levels: options.levels,
+            to: nodeColumns.length - 1, // Height of the tree
+            defaults: {
+                borderColor: options.borderColor,
+                borderRadius: options.borderRadius, // organization series
+                borderWidth: options.borderWidth,
+                color: series.color,
+                colorByPoint: options.colorByPoint,
+                // NOTE: if support for allowTraversingTree is added, then
+                // levelIsConstant should be optional.
+                levelIsConstant: true,
+                linkColor: options.linkColor, // organization series
+                linkLineWidth: options.linkLineWidth, // organization series
+                linkOpacity: options.linkOpacity,
+                states: options.states
+            }
+        });
+
+        this.layoutAlgorythm.calculatePositions(series);
+        series.layoutModifier = this.getLayoutModifiers();
+
+        // First translate all nodes so we can use them when drawing links
+        this.nodes.forEach((node): void => {
+            // To Change - wasVisited set to false to enable next generatePoints
```

Comment:
do we need this comment? Maybe we may set the `wasVisited `in point constructor and reset it in the update?

Reply:
Moved this property to translateNode.